### PR TITLE
WIN: Use copy instead of rsync for copying in copyfromremote

### PR DESCRIPTION
**Pull request by TamerGezici**
This is unrelated to the symlink project, so I'm pushing it seperately for clearance.

### CODE REVIEW

**Comment by TamerGezici** [comment]
Is the CI failure related to this? 

https://github.com/automaticanalysis/automaticanalysis/pull/364#issuecomment-1325652301

**Comment by tiborauer** [comment]
Yes

**Comment by tiborauer** [comment]
I was thinking... If we assume `~isempty(host)` indicated local copy, then the section could be simplified (and made more platform-independent) by
``` matlab
        if ~isempty(host)
            cmd = sprintf('rsync -t %s:''%s'' %s',host,src,dest);
            [s, w]=aas_shell(cmd,allow404,~allow404);
            if (s==0)
                if vargs.verbose
                    aas_log(aap,false,sprintf('Retrieved %s from %s',src,host),'m');
                end
                break;
            end;
            if (allow404 && ~isempty(strfind(w,'No such file or directory')))
                break;
            end;
            if (~isempty(strfind(w,'Connection refused')))
                aas_log(aap,false,sprintf('Connection refused in transfer, retry in %d s',retry));
                pause(retry)
            else
                break;
            end;
        else
            copyfile(src, dest);
        end     
```

File: aa_engine/aas_copyfromremote.m [hunk 1]
```diff
@@ -99,13 +99,28 @@
...
+                break;
+            end;
+            if (~isempty(strfind(w,'Connection refused')))
+                aas_log(aap,false,sprintf('Connection refused in transfer, retry in %d s',retry));
+                pause(retry)
+            else
+                break;
+            end;
+        end     
         if (s==0)
```
[the remark this answers is not in the document]
Comment (tiborauer) [reply aa_engine/aas_copyfromremote.m]: @TamerGezici , could you amend this?

**Review by tiborauer — APPROVED** [review]

File: aa_engine/aas_copyfromremote.m (hunk 1 above)
[the remark this answers is not in the document]
Comment (TamerGezici) [reply aa_engine/aas_copyfromremote.m]: I just did that before you posted, it should be fine now, isn't it?

**Comment by TamerGezici** [comment]
There may be a problem with this code. I will investigate and get back to you! Please don't merge yet.

**Comment by TamerGezici** [comment]
I think MATLAB inbuilt copyfile works different than Windows built `copy `command.

I ran into an error:

```
 retrieve stream epi [checksum match, not recopied] from F:r-pr-3Error using copyfile
No matching files named
'F:\fr-pr-3\B_block\studyB_prep\aamod_realign_00001\sub-12\stream_realignment_parameter_outputfrom_aamod_realign_00001.txt' were found.

Error in aas_copyfromremote (line 103)
            copyfile(src, dest);

Error in aas_retrieve_inputs_part1 (line 109)
            aap=aas_copyfromremote(aap,inputstream.host,remoteoutputstreamdesc,outputstreamdesc,'allow404',1,'allowcache',inputstream.allowcache, 'verbose', 1);

Error in aa_doprocessing_onetask (line 173)
                    [gotinputs, streamfiles{inpind}]=aas_retrieve_inputs_part1(aap,inp,allinputs,deps);

Error in aaq_localsingle/runall (line 15)
                aa_doprocessing_onetask(obj.aap,job.task,job.k,job.indices);

Error in aa_doprocessing (line 361)
            taskqueue.runall(dontcloseexistingworkers, false);

Error in studyB_MVPA_sr_rulebased_positionwise (line 49)
aa_doprocessing(aap);
```

Do pay attention that `stream_realignment_parameter_outputfrom_aamod_realign_00001.txt` file does exist, but it is under the directory for the session:

`F:\fr-pr-3\B_block\studyB_prep\aamod_realign_00001\sub-12\discrB_run-01`

and not under this folder:

`F:\fr-pr-3\B_block\studyB_prep\aamod_realign_00001\sub-12\`

When I replace the code `copyfile()` with 

```
cmd = sprintf('copy %s %s',src,dest);
[s, w]=aas_shell(cmd,allow404,~allow404);
```

The code will execute with no errors. And the file is also copied.

**Comment by tiborauer** [comment]
What are `src` and `dest`?

**Comment by TamerGezici** [comment]
```
K>> src

src =

    'F:\fr-pr-3\B_block\studyB_prep\aamod_realign_00001\sub-01\stream_realignment_parameter_outputfrom_aamod_realign_00001.txt'

K>> dest

dest =

    'F:\fr-pr-3\B_block\GLM_results\MVPA_sr_rulebased_acrossruns\aamod_firstlevel_model_00001\sub-01\stream_realignment_parameter_remoteoutputfrom__aamod_realign_00001.txt'
```

I think something happens in line 109 of aas_retrieve_inputs_part1.m which makes it search for the wrong directory. Maybe later on a proper call is made using the correct directory leading it to work. But when you do copyfile() and it can't find the directory, it throws an error (which maybe means copy command doesnt care about incorrect paths)

**Comment by tiborauer** [comment]
I think you are on a good lead. I suggest using debugging with a line-by-line execution. It may help clarifying the case.

**Comment by TamerGezici** [comment]
Ok, I can confirm what I described above is indeed what's happening.

I am using a different study as I currently do not have access to those files -- but the idea is same.

----------

So, I added these 3 lines under here to debug.

https://github.com/automaticanalysis/automaticanalysis/blob/1094ed7ac978ef2eb9f066c8eed41ed828b578fa/aa_engine/aas_retrieve_inputs_part1.m#L109

```
disp('\n')
disp(remoteoutputstreamdesc);
disp(outputstreamdesc);
```

So the first directory is the remote directory, and the other directory is the target directory.

Initially, it runs the function for these directories (however, the file doesn't exist under that directory) 
` hence the error: No matching files named`

```
D:\studies\analysis\patient_studies\sub-10-gogus\memorylocalizer_prep\aamod_realign_00001\sub-10\stream_realignment_parameter_outputfrom_aamod_realign_00001.txt
D:\studies\analysis\patient_studies\sub-10-gogus\github_test\aamod_firstlevel_model_00001\sub-10\stream_realignment_parameter_remoteoutputfrom__aamod_realign_00001.txt
```

Right after the call for the directories above, the next call is this, which is the correct directory:

```
D:\studies\analysis\patient_studies\sub-10-gogus\memorylocalizer_prep\aamod_realign_00001\sub-10\memorylocalizer_run-01\stream_realignment_parameter_outputfrom_aamod_realign_00001.txt
D:\studies\analysis\patient_studies\sub-10-gogus\github_test\aamod_firstlevel_model_00001\sub-10\memorylocalizer_run-01\stream_realignment_parameter_remoteoutputfrom__aamod_realign_00001.txt
```

The error will not happen if you use `copy `command but `copyfile()` doesn't like it when you try copying a file which doesn't exist, it seems!

**Comment by tiborauer** [comment]
Good approach! However, I mean using MATLAB's built-in debugging mode with breakpoints. So you can see all variables and the order of function calls.

My educated guess is that the destination is specified in an iterative approach. Calling the aas_shell does not crash and allows the next iteration, while copyfile crashes.

**Comment by TamerGezici** [comment]
Yes, that's what I've been doing, using break points but also disp helps with seeing the values of variables in loops.

Anyway, the issue seems to be related to the "domain" of this module.

https://github.com/automaticanalysis/automaticanalysis/blob/962c336b493f8acc0bad8845bc52a73c3285a2a7/aa_engine/aas_retrieve_inputs_part1.m#L11

The "domain" variable here is set to "subject" initially for realignment. And at the next call, it is set as "session" which is the correct one (so it works fine)

Down the line, here, the wrong path is assigned because we are looking for it at the wrong domain:

https://github.com/automaticanalysis/automaticanalysis/blob/962c336b493f8acc0bad8845bc52a73c3285a2a7/aa_engine/aas_retrieve_inputs_part1.m#L84

In the module header, I can see that here it is set as "subject":

https://github.com/automaticanalysis/automaticanalysis/blob/962c336b493f8acc0bad8845bc52a73c3285a2a7/aa_modules/aamod_realign.xml#L4

However, for slicetiming it is "session" and it works fine.

https://github.com/automaticanalysis/automaticanalysis/blob/962c336b493f8acc0bad8845bc52a73c3285a2a7/aa_modules/aamod_slicetiming.xml#L4

I wonder why realignment is in subject domain...

**Comment by tiborauer** [comment]
First, realigment MUST be in the _subject_ domain because there may be motion between the sessions, which MUST also be corrected.
In theory, you could run realignment for each session separately, but then you MUST coregister the session to each other or (separately) to the structural. The former would be equivalent with the realignement in the subject domain, while the latter would include more between-modality registration, which is less accurate than the within modality one.

Second, I am not surprised that the issue is domain-specific because if the domains match then only one iteration is needed. If not, then more iterations are needed, which would crash copyfile as I described earlier.  

I think the best solution would be to either 
- (simpler) put the copyfile into a try-catch block to prevent crashing and provide similar flags (variables s and w) as the aas_shell-copy solution. 
- (more elegant) run some checks (domain or folder) before the copy and skip unnecessary iteration(s).

**Comment by TamerGezici** [comment]
Thanks a lot for the explanation. It makes sense now.

What about moving this line

https://github.com/automaticanalysis/automaticanalysis/blob/962c336b493f8acc0bad8845bc52a73c3285a2a7/aa_engine/aas_retrieve_inputs_part1.m#L109

inside this if block?

https://github.com/automaticanalysis/automaticanalysis/blob/962c336b493f8acc0bad8845bc52a73c3285a2a7/aa_engine/aas_retrieve_inputs_part1.m#L111

**Comment by TamerGezici** [comment]
Okay, nevermind... This is for something else, it won't work.

**Comment by TamerGezici** [comment]
How about adding this:

```
if ~isfile(remoteoutputstreamdesc)
    continue;
end;
```
directly above: 

https://github.com/automaticanalysis/automaticanalysis/blob/962c336b493f8acc0bad8845bc52a73c3285a2a7/aa_engine/aas_retrieve_inputs_part1.m#L109

**Comment by tiborauer** [comment]
It seems like a good approach to me. Can you try it out?

**Comment by TamerGezici** [comment]
Yes, this seems to work. I tested it with a remote connected pipeline to do GLM + contrasts though. Anything else you think I should check?

**Comment by tiborauer** [comment]
Great news! Add it to the PR and see how the CI cope with it!

**Comment by TamerGezici** [comment]
Btw, I noticed something peculiar.

```
if isempty(host)
     copyfile(src, dest);
     if vargs.verbose
          aas_log(aap,false,sprintf('Retrieved %s from %s',src,host),'m');
    end
else
```

Adding this logging line here slows down the copying process vastly. When logging is off, data is transferred at speeds of 70-80 files per second, whereas when logging is on, it is 10 items per second.

**Comment by tiborauer** [comment]
I am not surprised. Printing text outputs usually slows down the execution. Besides, aas_log adds a bit (useful) convolution to texting. If it bothers you, set verbosity in aap.options.verbose (?).

**Comment by TamerGezici** [comment]
> aap.options.verbose

The verbosity option is off by default for calls using aas_copyfromremote in aas_retrieve_inputs_part2. It doesn't use that variable to set it I think. But it's good to know, thanks!

**Comment by TamerGezici** [comment]
I can't see the error in the CI, can you?

```
  ================================================================================
  Error occurred in testUseCases/use_case_test(testname=aatest_ds000114_AROMA_m) and it did not run to completion.
      ---------
      Error ID:
      ---------
      'aa:internal'
      --------------
      Error Details:
      --------------
      Error using aas_log (line 67)
      aa error:
      see above (or set aap.options.verbose)
```

**Comment by tiborauer** [comment]
It seems to be a python issue in ICA-AROMA. If you scroll up to the standard output. I am not sure what the cause is because it ran without error 3 days ago. I do not think it is due to your PR, though.

**Comment by TamerGezici** [comment]
To rule out the possibility that it somehow is related to my last commit, https://github.com/automaticanalysis/automaticanalysis/pull/365/commits/1177aa3dc9a593505330339c7f452218fdb1c347

I will just commit this to a seperate branch and make a PR.

**Comment by TamerGezici** [comment]
Seems like it is unrelated to my PR, indeed.

**Comment by tiborauer** [comment]
I am rerunning a previously finished test to confirm.

**Comment by TamerGezici** [comment]
> I am rerunning a previously finished test to confirm.

Did it error as well?

**Comment by tiborauer** [comment]
I am afraid it did, which means it has nothing to do with you PR but requires further investigation. It will be a can of worms/pythons. :)

**Comment by TamerGezici** [comment]
> I am afraid it did, which means it has nothing to do with you PR but requires further investigation. It will be a can of worms/pythons. :)

I see :) anything I can do? Or for this PR, is it fine to merge as it is? As there are some other PR's I wanted to create.

**Comment by TamerGezici** [comment]
By the way why is the `aas_shell` `quiet `parameter used with the `allow404 `variable? 

https://github.com/automaticanalysis/automaticanalysis/blob/962c336b493f8acc0bad8845bc52a73c3285a2a7/aa_engine/aas_copyfromremote.m#L108

Because this variable determines whether the command will be printed using `aas_log`

https://github.com/automaticanalysis/automaticanalysis/blob/962c336b493f8acc0bad8845bc52a73c3285a2a7/aa_engine/aas_shell.m#L4

Shouldn't it be a setting determined by the user such as verbose? 

So wouldn't using `aap.options.verbose` be better instead of using `allow404`?

**Comment by tiborauer** [comment]
You made a point. I suggest setting the defaults https://github.com/automaticanalysis/automaticanalysis/blob/962c336b493f8acc0bad8845bc52a73c3285a2a7/aa_engine/aas_copyfromremote.m#L18-L22
according to `aap.options.verbose`.

**Comment by TamerGezici** [comment]
@tiborauer what do you think of the last commit?

**Comment by tiborauer** [comment]
- `aap.options.verbose` can be [0-2], so you may want to explicitly convert it to logical (i.e., 0 -> false, [1 2] -> true)
- https://github.com/automaticanalysis/automaticanalysis/blob/da82888cda8bddcff452760054da87e1d341b243/aa_engine/aas_copyfromremote.m#L110 I would use `vargs.verbose` instead of `aap.options.verbose` as a parameter.

**Comment by TamerGezici** [comment]
Sure.

in `aas_shell `I noticed the parameter should have been the logical NOT of verbose, because `quiet `needs to be 1 so that it is not verbose, hence the negation operator.

I also noticed `aaq.directory_conventions.allowremotecache` was not a valid variable for me, which needed to be `aap`.... instead.

I also added "2" to the vdefaults, since the previous version only wrote [0, 1]

**Comment by tiborauer** [comment]
I am sorry for the confusion but using a `logical` conversion ensures that the range stays in the expected `[0 1]`, so you should not add "2" to the documentation.

**Comment by tiborauer** [comment]
Can you explain your motivation behind these changes? https://github.com/automaticanalysis/automaticanalysis/blob/592b8a5ce86ee395bf7d6c564cb0fc26f187234c/aa_engine/aas_retrieve_inputs_part1.m#L108-L110
My primary issue is that it does not generates any feedback to the user, which means some (expected) streams may by left our without a warning potentially producing a crash later or (in the worst case) unexpected output.

**Comment by TamerGezici** [comment]
> Can you explain your motivation behind these changes?
> 
> https://github.com/automaticanalysis/automaticanalysis/blob/592b8a5ce86ee395bf7d6c564cb0fc26f187234c/aa_engine/aas_retrieve_inputs_part1.m#L108-L110
> 
> 
> My primary issue is that it does not generates any feedback to the user, which means some (expected) streams may by left our without a warning potentially producing a crash later or (in the worst case) unexpected output.

I see, that is fair. I did it according to your request here (the last bit) https://github.com/automaticanalysis/automaticanalysis/pull/365#issuecomment-1362041941

What kind of a warning shall I give? Something like "files for _this_ stream are not available, please check it out" where _this_ is an appropriate variable that shows which stream to look for.

**Comment by tiborauer** [comment]
My bad! This is a very long thread, and I was unsure why you have done it. You do not need to provide any feedback; however, you may want to add a comment to the code to avoid further annoying questions. :) You could write something like this: "No stream has been found in this domain. Will check further (if any more left)."

**Comment by TamerGezici** [comment]
> My bad! This is a very long thread, and I was unsure why you have done it. You do not need to provide any feedback; however, you may want to add a comment to the code to avoid further annoying questions. :) You could write something like this: "No stream has been found in this domain. Will check further (if any more left)."

No worries! Amended.

**Pull request merged**